syntheseus/search/analysis/starting_molecule_match.py

```diff
@@ -0,0 +1,160 @@
+
+from __future__ import annotations
+
+import itertools
+import typing
+from collections.abc import Iterable
+
+from syntheseus.search.chem import Molecule
+from syntheseus.search.graph.and_or import ANDOR_NODE, AndOrGraph, OrNode
+
+T = typing.TypeVar("T")
+
+
+def is_route_with_starting_mols(
+    graph: AndOrGraph,
+    starting_mols: set[Molecule],
+    forbidden_nodes: typing.Optional[set[ANDOR_NODE]] = None,
+) -> bool:
+    """Checks whether there is a route in graph matching the starting mols."""
+    forbidden_nodes = forbidden_nodes or set()
+    return _is_route_with_starting_mols(
+        graph,
+        graph.root_node,
+        starting_mols,
+        forbidden_nodes,
+        _starting_mols_under_each_node(graph),
+    )
+
+
+def split_into_subsets(A: list[T], k: int) -> Iterable[list[list[T]]]:
+    """
+    Enumerate all possible ways to create k subsets from a list A such that none of the k subsets are empty,
+    the union of the sets is A, and the order of the subsets *does* matter.
+
+    This function is easiest to explain by example:
+
+    A single partition is just the list A itself.
+
+    >>> list(split_into_subsets([1, 2], 1))
+    >>> [[[1, 2]]]
+
+    For k=2, there are 4 possible subsets meaning 16 pairs of subsets,
+    but only 8 of them have non-empty subsets which include every element once.
+
+    >>> list(split_into_subsets([1, 2,], 2))
+    >>> [[[1], [2]], [[1], [1,2]], [[2], [1]], [[2], [1,2]], [[1,2], [1]], [[1,2], [2]]]
+    >>> [[[1], [2]], [[1], [1, 2]], [[2], [1]], [[2], [1, 2]], [[1, 2], [1]], [[1, 2], [2]], [[1, 2], [1, 2]]]
+
+    The implementation just uses itertools.combinations and itertools.products to enumerate all possible partions,
+    and simply rejects those which do not sum up to the entire set.
+    It is not very efficient for large A or large k, so use with caution.
+
+    NOTE: the efficiency of this method could definitely be improved later.
+    """
+
+    # Check 1: elements of list are unique
+    assert len(set(A)) == len(A)
+
+    # Check 2: k is valid
+    assert k >= 1
+
+    # Base case: list is empty
+    if len(A) == 0:
+        return
+
+    # Iterate through all subsets
+    power_set_non_empty = itertools.chain.from_iterable(
+        itertools.combinations(A, r) for r in range(1, len(A) + 1)
+    )
+    for subsets in itertools.product(power_set_non_empty, repeat=k):
+        if set(itertools.chain.from_iterable(subsets)) == set(A):
+            yield [list(s) for s in subsets]
+
+
+def _starting_mols_under_each_node(graph: AndOrGraph) -> dict[ANDOR_NODE, set[Molecule]]:
+    """Get set of molecules reachable under each node in the graph."""
+
+    # Initialize to empty sets, except for nodes with purchasable mols
+    node_to_mols: dict[ANDOR_NODE, set[Molecule]] = {n: set() for n in graph.nodes()}
+    for n in graph.nodes():
+        if isinstance(n, OrNode):
+            node_to_mols[n].add(n.mol)
```

Comment:
The comment says "purchasable", but is this reflected in the code?

Reply:
Good catch: I used to check whether a molecule is purchasable but removed it to make the method more general (i.e. the nodes marked purchasable in the graph don't need to be the starting molecules). I will fix this by changing the comment.